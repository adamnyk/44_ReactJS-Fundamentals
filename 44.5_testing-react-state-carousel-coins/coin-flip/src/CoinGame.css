.CoinGame{
    text-align: center;
}
.CoinGame-coin {
    width: 250px;
	height: 250px;
}

.CoinGame-count {
    margin: 10px;
}