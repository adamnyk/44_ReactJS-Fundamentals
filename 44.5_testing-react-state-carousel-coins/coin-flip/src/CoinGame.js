import React, { useState } from "react";
import Coin from "./Coin";
import "./CoinGame.css";

const CoinGame = (props) => {
	const [count, setCount] = useState({ heads: 0, tails: 0, total: 0 });
	const [coinSide, setCoinSide] = useState();

	function flipCoin() {
		const side = Math.floor(Math.random() * 2);
		setCoinSide(side);
		const countCopy = { ...count };
		countCopy.total++;
		side === 0 ? countCopy.heads++ : countCopy.tails++;
		setCount(countCopy);
	}

	const coin =
		coinSide !== undefined ? (
			<Coin {...props.coinImages[coinSide]} />
		) : (
			<div className="CoinGame-coin" />
		);

	return (
		<div className="CoinGame">
			{coin}
			<button onClick={flipCoin} data-testid="flip-button">
				Flip Coin!
			</button>

			<div className="CoinGame-count">
				<div>Heads: {count.heads}</div>
				<div>Tails: {count.tails}</div>
				<div>Total flips: {count.total}</div>
			</div>
		</div>
	);
};

CoinGame.defaultProps = {
	coinImages: [
		{
			src: "https://i.ebayimg.com/images/g/xtcAAOSwLwBaZigS/s-l500.jpg",
			alt: "heads",
		},
		{
			src: "https://m.media-amazon.com/images/I/51bcZy+HZpL._AC_UF894,1000_QL80_.jpg",
			alt: "tails",
		},
	],
};

export default CoinGame;
