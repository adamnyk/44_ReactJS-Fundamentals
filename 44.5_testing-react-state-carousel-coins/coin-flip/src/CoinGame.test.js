import React from "react";
import { render, fireEvent, queryByTestId } from "@testing-library/react";
import CoinGame from "./CoinGame";

// Mocks Math.random to return a specific first and second value when run
beforeEach(function () {
	jest
		.spyOn(Math, "random")
		.mockReturnValueOnce(0.25) // first call
		.mockReturnValueOnce(0.75); // second call then to default
});

afterEach(function () {
	Math.random.mockRestore();
});

// smoke test
it("renders iwthout crashing", () => {
	render(<CoinGame />);
});

// snapshot test
it("matches snapshot", () => {
	const { asFragment } = render(<CoinGame />);
	expect(asFragment()).toMatchSnapshot();
});

// testing elements
it("should not render the coin on start", function () {
	const { queryByAltText } = render(<CoinGame />);

	// expect the first image to show, but not the second
	expect(queryByAltText("heads")).not.toBeInTheDocument();
	expect(queryByAltText("tails")).not.toBeInTheDocument();
});

it("should show the coin when pressing the flip button", function () {
	const { queryByAltText, queryByTestId } = render(<CoinGame />);

	const flipButton = queryByTestId("flip-button");
	fireEvent.click(flipButton);

	// expect the coin to show, heads, when pressing the flip button
	expect(queryByAltText("heads")).toBeInTheDocument();

	// expect the coin to show, tails, when pressing the flip button
	fireEvent.click(flipButton);
	expect(queryByAltText("tails")).toBeInTheDocument();
});

it("should have all counters start at 0", function () {
	const { queryByText, queryByAltText } = render(<CoinGame />);

	expect(queryByAltText("tails")).not.toBeInTheDocument();
	expect(queryByAltText("heads")).not.toBeInTheDocument();

	// counters should all be 0 at start
	expect(queryByText("Heads: 0")).toBeInTheDocument();
	expect(queryByText("Tails: 0")).toBeInTheDocument();
	expect(queryByText("Total flips: 0")).toBeInTheDocument();
});

it("should count flips correctly", function () {
	const { queryByText, queryByAltText, queryByTestId } = render(<CoinGame />);

	// expect to count heads correctly
	const flipButton = queryByTestId("flip-button");
	fireEvent.click(flipButton);
	expect(queryByAltText("heads")).toBeInTheDocument();
	expect(queryByAltText("tails")).not.toBeInTheDocument();

	expect(queryByText("Heads: 1")).toBeInTheDocument();
	expect(queryByText("Tails: 0")).toBeInTheDocument();
	expect(queryByText("Total flips: 1")).toBeInTheDocument();

	// expect the coin to show, tails, when pressing the flip button
	fireEvent.click(flipButton);
	expect(queryByAltText("tails")).toBeInTheDocument();
	expect(queryByAltText("heads")).not.toBeInTheDocument();

	expect(queryByText("Heads: 1")).toBeInTheDocument();
	expect(queryByText("Tails: 1")).toBeInTheDocument();
	expect(queryByText("Total flips: 2")).toBeInTheDocument();
});
