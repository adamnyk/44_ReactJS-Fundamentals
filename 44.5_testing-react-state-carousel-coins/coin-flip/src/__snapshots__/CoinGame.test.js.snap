// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinGame"
  >
    <div
      class="CoinGame-coin"
    />
    <button
      data-testid="flip-button"
    >
      Flip Coin!
    </button>
    <div
      class="CoinGame-count"
    >
      <div>
        Heads: 0
      </div>
      <div>
        Tails: 0
      </div>
      <div>
        Total flips: 0
      </div>
    </div>
  </div>
</DocumentFragment>
`;
